### PR TITLE
Add ability to create template from VM for vSphere post-processor

### LINKED ISSUE #3035: vSphere post-processor mark as template
Is it possible to have the vSphere provisioner automatically mark the uploaded VM as a template?

### LINKED ISSUE #3809: vsphere post-processor should generate a template (not a VM)
```feature
Scenario: vsphere post-processor should generate a template

In order to produce a usable artifact
A packer user
wants to generate a vsphere template

Given I have the vsphere post-processor enabled
When and I run packer build 
Then I should see and template generated
And I should not see a virtual machine generated
```

### PR DESCRIPTION
**Pull request by samdunne**
Closes #3809
Closes #3035

### CODE REVIEW

**Comment by samdunne** [comment]
I'm not sure why the build fails on travis. The context package is vendored from what I can see. Ideas?

File: post-processor/vsphere/post-processor.go [hunk 1]
```diff
@@ -136,6 +140,43 @@ func (p *PostProcessor) PostProcess(ui packer.Ui, artifact packer.Artifact) (pac
...
+		sdkURL, err := url.Parse(fmt.Sprintf("https://%s:%s@%s/sdk",
+			url.QueryEscape(p.config.Username),
+			url.QueryEscape(p.config.Password),
+			p.config.Host))
+
+		if err != nil {
+			return nil, false, fmt.Errorf("Failed: %s\n", err)
+		}
+
+		client, err := govmomi.NewClient(context.TODO(), sdkURL, true)
```
Comment (rickard-von-essen) [inline post-processor/vsphere/post-processor.go]: TODO

File: post-processor/vsphere/post-processor.go [hunk 2]
```diff
@@ -136,6 +140,43 @@ func (p *PostProcessor) PostProcess(ui packer.Ui, artifact packer.Artifact) (pac
...
+
+		if err != nil {
+			return nil, false, fmt.Errorf("Failed: %s\n", err)
+		}
+
+		finder := find.NewFinder(client.Client, false)
+		datacenter, err := finder.DefaultDatacenter(context.TODO())
```
Comment (rickard-von-essen) [inline post-processor/vsphere/post-processor.go]: TODO

File: post-processor/vsphere/post-processor.go [hunk 3]
```diff
@@ -136,6 +140,43 @@ func (p *PostProcessor) PostProcess(ui packer.Ui, artifact packer.Artifact) (pac
...
+		finder.SetDatacenter(datacenter)
+		if err != nil {
+			return nil, false, fmt.Errorf("Failed: %s\n", err)
+		}
+
+		vm, err := finder.VirtualMachine(context.TODO(), p.config.VMName)
```
Comment (rickard-von-essen) [inline post-processor/vsphere/post-processor.go]: TODO

File: post-processor/vsphere/post-processor.go [hunk 4]
```diff
@@ -136,6 +140,43 @@ func (p *PostProcessor) PostProcess(ui packer.Ui, artifact packer.Artifact) (pac
...
+
+		if err != nil {
+			return nil, false, fmt.Errorf("Failed: %s\n", err)
+		}
+
+		ui.Message(fmt.Sprintf("Marking as template %s", p.config.VMName))
+		err = vm.MarkAsTemplate(context.TODO())
```
Comment (rickard-von-essen) [inline post-processor/vsphere/post-processor.go]: TODO

Comment (samdunne) [inline post-processor/vsphere/post-processor.go]: Is there an issue with using TODO? I'm reasonably new to Go 😂

Comment (rickard-von-essen) [inline post-processor/vsphere/post-processor.go]: `context.TODO()` might actually be fine but I would default to `context.Background()`.

Comment (samdunne) [inline post-processor/vsphere/post-processor.go]: 👍 I shall make the change and test

**Comment by samdunne** [comment]
@rickard-von-essen appveyor is building on Go 1.7.1 and Travis is building on 1.6. I upgraded the travis file and Vagrant file to use 1.7.1 because it fixes a bug with the new context package I am importing. 

Do you want me to split the change to a separate PR?

**Comment by samdunne** [comment]
Any feedback on this?

**Comment by rickard-von-essen** [comment]
@mwhooker I'm a bit uncomfortable updating the Go version to 1.7.1 on Travis. Do we have a strategy for this? If we should do that change I think it deservers it's own PR.

**Comment by rickard-von-essen** [comment]
And I think it would be better if the option would be `template` not `convert_to_template`.

**Comment by samdunne** [comment]
I'm happy to split the 1.7.1 change into a seperate and I'm not precious about the name of the variable. Whatever people think is best works

**Comment by mwhooker** [comment]
Hey, sorry to let this sit. There's a couple of trivial conflicts which should be resolved, then happy to merge.

A few things, though:

- Needs documenting. See website/source/docs/post-processors/vsphere.html.md
- `convert_to_template` seems fine, but this is not my area of expertise. The code seems to call `MarkAsTemplate`, so maybe something like that is more appropriate.
- I noticed vendor.json has this added

```
{
 +			"path": "context",
 +			"revision": ""
 +		},
```

That doesn't look right to me. We already have `golang.org/x/net/context` in the json, so I think that can come out. I know context is part of 1.7, but the net/context package is still available for BC. Since we have a few libraries that reference that path, I think it's best to keep for now.

**Comment by marema31** [comment]
@samdunne, I'd like to reuse some of your modification (integration of govmomi) to my current PR (#4882). Do you plan to update this part, or I can reuse it directly ?

**Comment by samdunne** [comment]
@marema31 I don't have the ability to test this with VMWare anymore. If you want to take the work directly I'll close this PR :)

**Comment by michaeltlombardi** [comment]
We're strongly interested in a vSphere builder at my org - I don't have any go experience, but I'm _glad_ to do some testing of either this or #4882 (cc @marema31) and provide detailed logs/notes. Also glad to help with documentation.

**Pull request closed without merging**